File: clients/csharp/Model/SoundProperties.cs
namespace AiCup22.Model
{
    /// <summary>
    /// Sound properties
    /// </summary>
    public struct SoundProperties
    {
        /// <summary>
        /// Name
        /// </summary>
        public string Name { get; set; }
        /// <summary>
        /// Distance from which the sound can be heard
        /// </summary>
        public double Distance { get; set; }
        /// <summary>
        /// Offset modifier
        /// </summary>
        public double Offset { get; set; }
    
        public SoundProperties(string name, double distance, double offset)
        {
            this.Name = name;
            this.Distance = distance;
            this.Offset = offset;
        }
    
        /// <summary> Read SoundProperties from reader </summary>
        public static SoundProperties ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new SoundProperties();
            result.Name = System.Text.Encoding.UTF8.GetString(reader.ReadBytes(reader.ReadInt32()));
            result.Distance = reader.ReadDouble();
            result.Offset = reader.ReadDouble();
            return result;
        }
    
        /// <summary> Write SoundProperties to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            var nameData = System.Text.Encoding.UTF8.GetBytes(Name);
            writer.Write(nameData.Length);
            writer.Write(nameData);
            writer.Write(Distance);
            writer.Write(Offset);
        }
    
        /// <summary> Get string representation of SoundProperties </summary>
        public override string ToString() {
            string stringResult = "SoundProperties { ";
            stringResult += "Name: ";
            stringResult += "\"" + Name + "\"";
            stringResult += ", ";
            stringResult += "Distance: ";
            stringResult += Distance.ToString();
            stringResult += ", ";
            stringResult += "Offset: ";
            stringResult += Offset.ToString();
            stringResult += " }";
            return stringResult;
        }
    }
}

File: clients/fsharp/Model/SoundProperties.fs
#nowarn "0058"

namespace AiCup22
namespace AiCup22.Model

open AiCup22

/// Sound properties
type SoundProperties = {
    /// Name
    Name: string;
    /// Distance from which the sound can be heard
    Distance: double;
    /// Offset modifier
    Offset: double;
} with

    /// Write SoundProperties to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        let nameData : byte[] = System.Text.Encoding.UTF8.GetBytes this.Name
        writer.Write nameData.Length
        writer.Write nameData
        writer.Write this.Distance
        writer.Write this.Offset
        ()

    /// Read SoundProperties from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Name = reader.ReadInt32() |> reader.ReadBytes |> System.Text.Encoding.UTF8.GetString
        Distance = reader.ReadDouble()
        Offset = reader.ReadDouble()
    }